### PR TITLE
Use single grammar for both stateless and stateful ALUs.

### PR DESCRIPTION
**Pull request by aatish17varma**

### CODE REVIEW

**Comment by anirudhSK** [comment]
Thanks for resolving these issues. Why is there a file temp/alu.g4? Is it used anywhere? Also in the last commit, it seems like it's taking forever to run some unit tests. Can we run those tests on cyclops to see what's going on?

**Comment by anirudhSK** [comment]
@mdw362 and @aatish17varma 

I ran the unit tests on cyclops. 19 of them fail. The most common error message seems to be:

AssertionError: Error:  type is declared as stateful and not 'stateless' for stateless ALU

Here's a test that causes this error:

<pre>
anirudh@cyclops:~/chipmunk$ python3 -m unittest tests.test_iterative_solver.IterativeSolverTest.test_times_two_hole_elimination
Iteration #1
ANTLR runtime and generated code versions disagree: 4.7.2!=4.5.1
ANTLR runtime and generated code versions disagree: 4.7.2!=4.5.1
F
======================================================================
FAIL: test_times_two_hole_elimination (tests.test_iterative_solver.IterativeSolverTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/anirudh/chipmunk/tests/test_iterative_solver.py", line 246, in test_times_two_hole_elimination
    '3', '3', '10', '--hole-elimination']),
  File "/home/anirudh/chipmunk/chipc/iterative_solver.py", line 167, in main
    additional_testcases=additional_testcases)
  File "/home/anirudh/chipmunk/chipc/compiler.py", line 135, in serial_codegen
    str(iter_cnt) + '.sk'))
  File "/home/anirudh/chipmunk/chipc/compiler.py", line 101, in single_codegen_run
    additional_testcases=additional_testcases)
  File "/home/anirudh/chipmunk/chipc/sketch_generator.py", line 267, in generate_sketch
    alu_definitions = self.generate_alus()
  File "/home/anirudh/chipmunk/chipc/sketch_generator.py", line 238, in generate_alus
    for k in range(0, self.num_phv_containers_)
  File "/home/anirudh/chipmunk/chipc/sketch_generator.py", line 92, in generate_stateless_alu
    stateless_alu_sketch_generator.visit(tree)
  File "/home/anirudh/.local/lib/python3.5/site-packages/antlr4/tree/Tree.py", line 34, in visit
    return tree.accept(self)
  File "/home/anirudh/chipmunk/chipc/aluParser.py", line 2493, in accept
    return visitor.visitAlu(self)
  File "/home/anirudh/chipmunk/chipc/stateless_alu_sketch_generator.py", line 38, in visitAlu
    self.visit(ctx.getChild(0, aluParser.State_indicatorContext))
  File "/home/anirudh/.local/lib/python3.5/site-packages/antlr4/tree/Tree.py", line 34, in visit
    return tree.accept(self)
  File "/home/anirudh/chipmunk/chipc/aluParser.py", line 507, in accept
    return visitor.visitState_indicator(self)
  File "/home/anirudh/chipmunk/chipc/stateless_alu_sketch_generator.py", line 73, in visitState_indicator
    ' and not \'stateless\' for stateless ALU '
AssertionError: Error:  type is declared as stateful and not 'stateless' for stateless ALU 

----------------------------------------------------------------------
Ran 1 test in 0.040s

FAILED (failures=1)
</pre>

**Comment by anirudhSK** [comment]
I fixed the assertion error from the command above, but the unit tests are taking much longer than the unit tests at the master branch. The number of hole bits in the generated sketch files also seem to be different from before.

**Comment by michaeldwong** [comment]
The test_simple_raw_succeeds_with_two_two_grid will never pass after the recent changes. The new structure of the stateless ALU functions is different from the old layout within the Sketch file in tests/data resulting in differing hole variables/values so the last assertion will always fail.

**Comment by anirudhSK** [comment]
@mdw362: Thanks for these changes. Unlike before, the unit tests run to completion on AppVeyor now, but 4 of them still fail. Two questions.

1. Can we check that the number of hole bits in all the generated sketches (belonging to the test suite) are the same before and after applying this pull request? You can determine the number of hole bits by searching for the string "Total number of hole bits is ..." This still only tells us that the number of hole bits in the generated sketch is the same as before---not that the generated sketches themselves are the same---but it's still a good sanity check.

2. I don't understand why "test_simple_raw_succeeds_with_two_two_grid will never pass." Could you explain this in more detail? If we need to change the reference output in the unit tests because the names of the holes have changed, we can do that.

**Comment by michaeldwong** [comment]
@anirudhSK
1. Okay I can try that and compare the hole bits once this build completes

2. The test_simple_raw_succeeds_with_two_two_grid function inside of tests/test_chipmunk.py asserts that the two dicts output_holes and expected_holes are equivalent. However, because of my changes in the structure of how stateless ALUs are used, the names of the holes are different. I've attached txt files showing the differences. By just eyeballing it, you can see that the hole bits are the same and it's just the names that are different (unless I missed something).
[output.txt](https://github.com/chipmunk-project/chipmunk/files/3413791/output.txt)
[expected.txt](https://github.com/chipmunk-project/chipmunk/files/3413792/expected.txt)

File: chipc/templates/router_data_path_sketch_synthesized_allocation.j2 [hunk 1]
```diff
@@ -35,11 +35,7 @@
     {% if stage_number == 0 %}
       // Read each PHV container from corresponding packet field.
       {% for field_number in range(num_fields_in_prog) %}
-        {% for container_number in range(num_phv_containers) %}
-          if (phv_config_{{field_number}}_{{container_number}} == 1){
-            input_0_{{container_number}} = state_and_packet.pkt_{{field_number}};
-          }
-        {% endfor %}
+        input_0_{{field_number}} = state_and_packet.pkt_{{field_number}};
```
Comment (anirudhSK) [inline chipc/templates/router_data_path_sketch_synthesized_allocation.j2]: Why did we change this? @XiangyuG: introduced this change in https://github.com/chipmunk-project/chipmunk/pull/151 for benchmarking. I think we need to revert this and the other change to router_data_path_sketch_synthesized_allocation.j2.

File: chipc/templates/router_data_path_sketch_synthesized_allocation.j2 [hunk 2]
```diff
@@ -152,11 +148,7 @@
 
   {% for field_number in range(num_fields_in_prog) %}
     // Write pkt_{{field_number}}
-    {% for container_number in range(num_phv_containers) %}
-      if (phv_config_{{field_number}}_{{container_number}} == 1){
-        state_and_packet.pkt_{{field_number}} = output_{{num_pipeline_stages - 1}}_{{container_number}};
-      }
-    {% endfor %}
+    state_and_packet.pkt_{{field_number}} = output_{{num_pipeline_stages - 1}}_{{field_number}};
```
Comment (anirudhSK) [inline chipc/templates/router_data_path_sketch_synthesized_allocation.j2]: Same comments as above.

File: example_alus/stateless_alus/stateless_alu_arith.alu [hunk 3]
```diff
@@ -0,0 +1,19 @@
+// Max value of opcode is 4
```
Comment (anirudhSK) [inline example_alus/stateless_alus/stateless_alu_arith.alu]: These numbers are 1 less than what they were previously. Also, the max value of opcode was used to determine the opcode bit width previously. Are we still doing that now?

File: chipc/templates/router_data_path_sketch_synthesized_allocation.j2 (hunk 2 above)
Comment (anirudhSK) [inline chipc/templates/router_data_path_sketch_synthesized_allocation.j2]: Why did we change this? @XiangyuG: introduced this change in #151 for benchmarking. I think we need to revert this and the other change to router_data_path_sketch_synthesized_allocation.j2.

File: chipc/templates/router_data_path_sketch_synthesized_allocation.j2 (hunk 1 above)
Comment (anirudhSK) [inline chipc/templates/router_data_path_sketch_synthesized_allocation.j2]: Why did we change this? @XiangyuG: introduced this change in https://github.com/chipmunk-project/chipmunk/pull/151 for benchmarking. I think we need to revert this and the other change to router_data_path_sketch_synthesized_allocation.j2.

File: example_alus/stateless_alus/stateless_alu_arith_rel.alu [hunk 4]
```diff
@@ -0,0 +1,35 @@
+// Max value of opcode is 12
```
Comment (anirudhSK) [inline example_alus/stateless_alus/stateless_alu_arith_rel.alu]: Same comment as before on max value of opcode.

File: example_alus/stateless_alus/stateless_alu_arith_rel_cond.alu [hunk 5]
```diff
@@ -0,0 +1,49 @@
+// Max value of opcode is 14
```
Comment (anirudhSK) [inline example_alus/stateless_alus/stateless_alu_arith_rel_cond.alu]: Same comment as before on max value of opcode.

File: example_alus/stateless_alus/stateless_alu_arith_rel_cond_bool.alu [hunk 6]
```diff
@@ -0,0 +1,58 @@
+// Max value of opcode is 19
```
Comment (anirudhSK) [inline example_alus/stateless_alus/stateless_alu_arith_rel_cond_bool.alu]: Same comment as before on max value of opcode.

File: chipc/sketch_generator.py [hunk 7]
```diff
@@ -81,70 +80,59 @@ def add_assert(self, assert_predicate):
...
+            stateless_alu_sketch_generator.num_packet_fields)
 
-        self.add_hole(self.sketch_name_ + '_' +
-                      alu_name + '_opcode', bit_of_opcode)
-        self.add_hole(self.sketch_name_ + '_' + alu_name + '_immediate', 2)
-        self.add_assert(
-            '(' + self.sketch_name_ + '_' + alu_name + '_opcode == 1)' + '|| ('
-            + self.sketch_name_ + '_' + alu_name + '_mux1_ctrl <= ' +
-            self.sketch_name_ + '_' + alu_name +
-            '_mux2_ctrl)')  # symmetry breaking for commutativity
```
Comment (anirudhSK) [inline chipc/sketch_generator.py:124]: Do we have these symmetry breaking asserts in the new code as well?

File: chipc/stateful_alu_sketch_generator.py [hunk 8]
```diff
@@ -1,242 +1,218 @@
...
-    def visitMux2(self, ctx):
-        self.mainFunction += self.stateful_alu_name + '_' + 'Mux2_' + str(
-            self.mux2Count) + '('
-        self.visit(ctx.getChild(0, stateful_aluParser.ExprContext))
-        self.mainFunction += ','
-        self.visit(ctx.getChild(1, stateful_aluParser.ExprContext))
-        self.mainFunction += ',' + 'Mux2_' + str(self.mux2Count) + ')'
-        self.generateMux2()
-        self.mux2Count += 1
+    def visitUpdates(self, ctx):
```
Comment (anirudhSK) [inline chipc/stateful_alu_sketch_generator.py:125]: Can we add the overrides annotation to all functions that we override?

**Review by anirudhSK — CHANGES_REQUESTED** [review]

**Comment by anirudhSK** [comment]
@XiangyuG and @taegyunkim: Can you look at this as well? It's a more invasive change, so would be good to take a closer look.

**Comment by taegyunkim** [comment]
I'll review this before dinner today.

File: chipc/sketch_generator.py [hunk 9]
```diff
@@ -81,70 +80,59 @@ def add_assert(self, assert_predicate):
...
-            mux1=self.sketch_name_ + '_' + alu_name + '_mux1',
-            mux2=self.sketch_name_ + '_' + alu_name + '_mux2',
-            mux3=self.sketch_name_ + '_' + alu_name + '_mux3')
-        # TODO: now fix # of mux to be 3 and will later make them flexible
-        # according to the number of mux used in stateless_alu
         mux_op_1 = self.generate_mux(
             len(potential_operands), alu_name + '_mux1')
         mux_op_2 = self.generate_mux(
             len(potential_operands), alu_name + '_mux2')
         mux_op_3 = self.generate_mux(
```
Comment (anirudhSK) [inline chipc/sketch_generator.py]: Can we determine the number of such operand muxes from the .alu file instead of hardcoding it to 3?

**Review by anirudhSK — CHANGES_REQUESTED** [review]

File: chipc/generator_tester.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+from aluLexer import aluLexer
+from aluParser import aluParser
+from antlr4 import CommonTokenStream
+from antlr4 import FileStream
+
+from chipc.alu.stateful_alu_sketch_generator import StatefulALUSketchGenerator
+
+
+def main(argv):
```
Comment (taegyunkim) [inline chipc/generator_tester.py]: What does this file do? What is the purpose of  this script?

If this is to test something, why not just write a unit test?

File: chipc/generator_tester.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    alu_file = argv[1]
```
Comment (taegyunkim) [inline chipc/generator_tester.py]: Consider using argparse package.

File: chipc/sketch_generator.py [hunk 12]
```diff
@@ -295,6 +283,7 @@ def generate_sketch(self, program_file, mode, additional_constraints=[],
             num_alus_per_stage=self.num_alus_per_stage_,
             num_phv_containers=self.num_phv_containers_,
             hole_definitions=self.hole_preamble_,
+
```
Comment (taegyunkim) [inline chipc/sketch_generator.py]: Remove unnecessary newline.

File: chipc/sketch_generator.py [hunk 13]
```diff
@@ -81,70 +80,59 @@ def add_assert(self, assert_predicate):
...
+        return (mux_op_1 +
+                mux_op_2 +
+                mux_op_3 +
+                stateless_alu_sketch_generator.helperFunctionStrings +
+                stateless_alu_sketch_generator.mainFunction)
 
     # Generate Sketch code for a simple stateful alu (+,-,*,/)
     # Takes one state and one packet operand (or immediate operand) as inputs
     # Updates the state in place and returns the old value of the state
+
```
Comment (taegyunkim) [inline chipc/sketch_generator.py]: Remove unnecessary newline

File: chipc/templates/router_data_path_sketch_synthesized_allocation.j2 (hunk 1 above)
Comment (taegyunkim) [reply chipc/templates/router_data_path_sketch_synthesized_allocation.j2]: Agree with @anirudhSK

File: example_alus/stateless_alus/stateless_alu.alu [hunk 14]
```diff
@@ -0,0 +1,59 @@
+// Max value of opcode is 19
```
Comment (taegyunkim) [inline example_alus/stateless_alus/stateless_alu.alu]: Same goes to this file, as @anirudhSK commented.

**Review by taegyunkim — CHANGES_REQUESTED** [review]

**Comment by anirudhSK** [comment]
Currently, the AppVeyor tests seem to be failing because of the SKETCH bug here: https://lists.csail.mit.edu/pipermail/sketchusers/2019-July/000092.html @taegyunkim: is there a way to work around the bug from our side before it's fixed by SKETCH?

**Review by anirudhSK — APPROVED** [review]

**Review by taegyunkim — APPROVED** [review]

**Pull request merged**